docs/proposals/auditing.md

```diff
@@ -0,0 +1,276 @@
+
+- `HttpHeaders`,
+- `RequestObject`,
+- `StorageObject`.
+
+In addition the policy action can contain a number of output backend dependent key/values e.g. to define JSON object paths which should be logged or excluded from logging. For portable policy actions, a number of key/values are standardised and ought to be supported by output backends.
+
+When the http request is processed, the request handler will close the audit event by filling in the http response. Then it will pass the event to the configured policy backend.
+
+**Note:** for service creation and deletion there is special REST code in the apiserver which takes care of service/node port (de)allocation and removal of endpoints on service deletion. Hence, these operations are not visible on the API layer and cannot be audit logged therefore. **No other resources** (with the exception of componentstatus which is not of interest here) **implement this kind of custom CRUD operations.**
+
+### Events
+
+```go
+package audit
+
+// LogLevel defines the amount of information logged for auditing
+type LogLevel int
+
+
+type Event struct {
+  Level LogLevel
+
+  // http header level (HeaderLogLevel and higher)
+  ID string
+  Timestamp time.Timestamp
+  IP string
+  Method string
+  URI string
+  User string
+  AsUser string
+  Namespace, Name string
+  GroupKind unversioned.GroupKind
+  Response int
+
+  // CRUD level (RequestLogLevel and higher)
+  RequestObject runtime.Unstructured // before admission
+
+  // Storage level (StorageLogLevel and higher)
+  OldObject runtime.Object
+  NewObject runtime.Object
+  Patch     []byte
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Policy
+
+```go
+package audit
+
+const (
+  // DontLogLevel means to not log at all
+  DontLogLevel LogLevel = iota
+
+  // HeaderLogLevel means to log only information from the http headers, not from the body
+  HeaderLogLevel
+
+  // RequestLogLevel adds the user request in the body
+  RequestLogLevel
+
+  // StorageLogLevel adds the old and the new object that is received from and sent to the storage layer
+  StorageLogLevel
+)
+
+type Action interface {
+  Level() LogLevel
+  Value(key string) string
+}
+
+type Policy interface {
+  func Action(kind, namespace, method, user string) Action
+}
+```
+
+TODO: how do the storage layer and the http handlers get access to the policy?
+
+### Policy Actions
+
+TODO: define what we need next to the audit level
+
+### Output Backend Interface
+
+```go
+package audit
+
+type OutputBackend interface {
+  Log(e *Event, a Action) error
+}
+```
+
+### Apiserver Command Line Flags
+
+```shell
+$ kube-apiserver --audit-output file:path=/var/log/apiserver-audit.log,rotate=1d,max=1024MB,format=json \
+  --audit-policy user=EndUser,kind=ReplicationController:output=diff,exclude=status
```

Comment:
I think this is likely to be pretty complex and expressive (particularly given pluggable backends).  Since we have configmaps in place today, is there a reason this shouldn't be restricted to a file ref for the entire `--audit-config`?